### ISSUE TITLE
Lint should warn about non-empty `return` in generator functions 

### ISSUE THREAD

**Issue by juesato** [PyCQA/pylint#7813]
### Current problem

The PEP (https://peps.python.org/pep-0255/#specification-return) says:

> A generator function can also contain return statements of the form:
>
> return
>
> Note that an expression_list is not allowed on return statements in the body of a generator

It would be nice if gpylint checked for cases of functions containing `yield` and non-empty `return`, and had a warning.

Otherwise, this can lead to surprising errors. For example, consider the following (which is an easy error to make):

```
import itertools
def repeat(ds, infinite=False):
  if infinite:
    return itertools.cycle(ds)
  for _ in range(2):
    for item in ds:
      yield item
```

If we run this method with `infinite=True`, it will return a length-0 iterator. (`return` should of course be replaced with `yield from`.)

### Desired solution

It would be nice if gpylint checked for cases of functions containing yield and non-empty return, and had a warning.

### Additional context

Apologies if this is a duplicate. I did a quick skim and didn't see anything.

**Comment by DanielNoord** [comment]
I actually use this behaviour at work to access a value that is being counted by the generator and then finally returned. Although it is indeed a bit hacky it seems like there are use cases for a pattern where `return` returns more than `None`.
Also note that `Generator` as a typing hint allows more than `None` as its third value.

**Comment by cdce8p** [comment]
> The PEP (https://peps.python.org/pep-0255/#specification-return) says:
> 
> > A generator function can also contain return statements of the form:
> > return
> > Note that an expression_list is not allowed on return statements in the body of a generator

That's updated. It changed in 3.3. The documentation now says
> In a generator function, the [return](https://docs.python.org/3/reference/simple_stmts.html#return) statement indicates that the generator is done and will cause [StopIteration](https://docs.python.org/3/library/exceptions.html#StopIteration) to be raised. The returned value (if any) is used as an argument to construct [StopIteration](https://docs.python.org/3/library/exceptions.html#StopIteration) and becomes the StopIteration.value attribute.

https://docs.python.org/3/reference/simple_stmts.html#the-return-statement

> Otherwise, this can lead to surprising errors.

I would recommend using type annotations and a type checker like `mypy` or `pyright` for that. Both will have no issue detecting this.

--
I recommend we close this one.

**Comment by juesato** [comment]
Ah, didn't see the Python 3 docs, thanks! Closing this thread.

> I would recommend using type annotations and a type checker like mypy or pyright for that. Both will have no issue detecting this.

OOC, do you have an example of how to type this? Both `pytype` and `mypy` are happy with this:

```
import itertools
import typing as tp

_T = tp.TypeVar('_T')

def repeat(
    ds: tp.Iterable[_T],
    infinite: bool = False) -> tp.Iterator[_T]:
  if infinite:
    return itertools.cycle(ds)
  for _ in range(2):
    for item in ds:
      yield item

repeat(range(3), infinite=True)
```

**Comment by cdce8p** [comment]
> > I would recommend using type annotations and a type checker like mypy or pyright for that. Both will have no issue detecting this.
> 
> OOC, do you have an example of how to type this? Both `pytype` and `mypy` are happy with this:

Try `Generator[_T, None, None]` instead. The last argument is the return type.
```py
import itertools
from typing import Generator, Iterable, TypeVar

_T = TypeVar('_T')

def repeat(
    ds: Iterable[_T],
    infinite: bool = False
) -> Generator[_T, None, None]:  # <-- Use Generator here
    if infinite:
        return itertools.cycle(ds)  # error: No return value expected  [return-value]
    for _ in range(2):
        for item in ds:
            yield item

repeat(range(3), infinite=True)
```
